test/unique_ptr_test.cpp

```diff
@@ -0,0 +1,256 @@
+const std::string LIBRARY_2 = "libclass_loader_TestPlugins2.so";
+
+using class_loader::ClassLoader;
+
+/*****************************************************************************/
+TEST(ClassLoaderUniquePtrTest, basicLoad)
+{
+  try
+  {
+    ClassLoader loader1(LIBRARY_1, false);
+    loader1.createUniqueInstance<Base>("Cat")->saySomething(); //See if lazy load works
+    SUCCEED();
+  }
+  catch(class_loader::ClassLoaderException& e)
+  {
+    FAIL() << "ClassLoaderException: " << e.what() << "\n";
+  }
+}
+
+/*****************************************************************************/
+TEST(ClassLoaderUniquePtrTest, correctLazyLoadUnload)
+{
+  try
+  {
+    ASSERT_FALSE(class_loader::class_loader_private::isLibraryLoadedByAnybody(LIBRARY_1));
+    ClassLoader loader1(LIBRARY_1, true);
+    ASSERT_FALSE(class_loader::class_loader_private::isLibraryLoadedByAnybody(LIBRARY_1));
+    ASSERT_FALSE(loader1.isLibraryLoaded());
+
+    {
+      ClassLoader::UniquePtr<Base> obj = loader1.createUniqueInstance<Base>("Cat");
+      ASSERT_TRUE(class_loader::class_loader_private::isLibraryLoadedByAnybody(LIBRARY_1));
+      ASSERT_TRUE(loader1.isLibraryLoaded());
+    }
+
+    //The library will unload automatically when the only plugin object left is destroyed
+    ASSERT_FALSE(class_loader::class_loader_private::isLibraryLoadedByAnybody(LIBRARY_1));
+    return;
+  }
+  catch(class_loader::ClassLoaderException& e)
+  {
+    FAIL() << "ClassLoaderException: " << e.what() << "\n";
+  }
+  catch(...)
+  {
+    FAIL() << "Unhandled exception";
+  }
+}
+
+/*****************************************************************************/
+
+TEST(ClassLoaderUniquePtrTest, nonExistentPlugin)
+{
+  ClassLoader loader1(LIBRARY_1, false);
+
+  try
+  {
+    ClassLoader::UniquePtr<Base> obj = loader1.createUniqueInstance<Base>("Bear");
+    if(obj == NULL)
+      FAIL() << "Null object being returned instead of exception thrown.";
+
+    obj->saySomething();
+  }
+  catch(const class_loader::CreateClassException& e)
+  {
+    SUCCEED();
+    return;
+  }
+  catch(...)
+  {
+    FAIL() << "Unknown exception caught.\n";
+  }
+
+  FAIL() << "Did not throw exception as expected.\n";
+}
+
+/*****************************************************************************/
+
+void wait(int seconds)
+{
+  boost::this_thread::sleep(boost::posix_time::seconds(seconds));
+}
+
+void run(ClassLoader* loader)
+{
+  std::vector<std::string> classes = loader->getAvailableClasses<Base>();
+  for(unsigned int c = 0; c < classes.size(); c++)
+  {
+    loader->createUniqueInstance<Base>(classes.at(c))->saySomething();
+  }
+}
+
+TEST(ClassLoaderUniquePtrTest, threadSafety)
+{
+  ClassLoader loader1(LIBRARY_1);
+  ASSERT_TRUE(loader1.isLibraryLoaded());
+
+  //Note: Hard to test thread safety to make sure memory isn't corrupted. The hope is this test is hard enough that once in a while it'll segfault or something if there's some implementation error.
```

Comment:
Note that I copied this from the `shared_ptr` test. It's somewhat double, since this test is actually testing if the deleter behaves correctly. Both the shared and unique pointers have the same deleter though, so unless the `unique_ptr` implementation is buggy (which is unlikely), the behaviour should be the same. Still, I figured it can't hurt to test this with the `unique_ptr` too.